### PR TITLE
[Refactor]Clean up the the camera code and implemented double click for orbit mode

### PR DESCRIPTION
**Pull request by nilscognite**
#### Type of change

Refactor 
#### Jira ticket :blue_book:
https://cognitedata.atlassian.net/browse/BND3D-2702
This PR is the beginning of making the **orbit mode!**. This code make it possible (this is default disabled for the moment), to double click at a object so it zooms to this object and set target in the center.

https://cognitedata.atlassian.net/browse/

## Description :pencil:

**Viewer.tsx:** Updated the new settings on the camera control: `changeCameraPositionOnDoubleClick`
Change the default on `changeCameraTargetOnClick `to false

**Image360Helper:** Reused `getPixelCoordinatesToNormalized`

**Cognite3DViewer:** Reused `getPixelCoordinatesToNormalized` , updated `modelIntersectionCallback `to get the clicked boundingBox box.

**viewer/packages/camera-manager/src/ComboControlsOptions.ts:**

I have done some cleanup in the `ComboControls `class without really changing anything:
* Move methods that did not use this function out, so they are outside the class.
* Move member fields that didn't change out as constants.
* Made the code homogeneous.
* Sort methods and member fields according to functionality and signature
* Moved `ComboControlsOptions `and some constant used by it to separate file
* Remove the member field `_firstPersonMode`.  This is not longer needed.
* Some minor changes to avoid repetitions.

**viewer/packages/camera-manager/src/DefaultCameraManager.ts**: 
* Cleanups, some renaming and moved some functions to make it more readable.
* Remove function `convertPixelCoordinatesToNormalized`
* Used clamp instead of combo of min....max functions. **Should be done other places in Reveal also!!!!**
* Implementeed and use the new `onDoubleClick `function. This is actually the main new functionality

**viewer/packages/camera-manager/src/types.ts**: Added a new option in two types. In `CameraManagerCallbackData `I addeed `pickedBoundingBox `to get the boundingbox of the object the mouse pick on.

**viewer/packages/tools/src/Measurement/MeasurementManager.ts**: Reused `getPixelCoordinatesToNormalized`. Only cleaned up the function using it. by moving decration closesr to the usage.

**viewer/packages/utilities/src/worldToViewport.ts:** Added a function. 

## TODO for removing tech depth in the future:

This is based on my limited work so far.

* Remove `pixelToNormalizedDeviceCoordinates `and use `getPixelCoordinatesToNormalized `instead.
* Go through all usage of min/max and replace by clamp.
* Split files so one type or class correspond to one file. Remove files like type.ts for instance and split them up.
* I also see that you break your own code standard by naming private member without the _ prefix.
* Be more aware of usage. Never copy a function. This is not reuse, but adding tech dept to the code base.

## How has this been tested? :mag:

Use the example app. Under camera controls, turn on the double click option. Do this on a real data, since it reads from CDF to find the bounding box.

## Checklist :ballot_box_with_check:

- [ ] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [ ] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/utilities/src/worldToViewport.ts [hunk 1]
```diff
@@ -71,5 +71,15 @@ export function worldToViewportCoordinates(
  * Converts a pixel coordinate to normalized device coordinate (in range [-1, 1])
  */
 export function pixelToNormalizedDeviceCoordinates(x: number, y: number, width: number, height: number): THREE.Vector2 {
-  return new THREE.Vector2((x / width) * 2 - 1, (y / height) * -2 + 1);
+  return new THREE.Vector2((x / width) * 2 - 1, -(y / height) * 2 + 1);
+}
+
+export function getPixelCoordinatesToNormalized(
```
Comment (haakonflatval-cognite) [inline viewer/packages/utilities/src/worldToViewport.ts:77]: It seems to me that this is mostly the same as the function above. Could we make use of `pixelToNormalizedDeviceCoordinates` by calling it from inside this function?

Comment (haakonflatval-cognite) [reply viewer/packages/utilities/src/worldToViewport.ts:77]: Also - the name is a bit confusing, I think I'd prefer something like `getNormalizedPixelCoordinates`instead.

Comment (nilscognite) [reply viewer/packages/utilities/src/worldToViewport.ts:77]: Yes. The first one would be removed in the next PR. Renaming the othewr ok

Comment (nilscognite) [reply viewer/packages/utilities/src/worldToViewport.ts]: 314413c300575b6f80c3c681266750399617bf9c

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM :+1:

File: viewer/packages/tools/src/Measurement/MeasurementManager.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { HtmlOverlayTool } from '../HtmlOverlay/HtmlOverlayTool';
 import { MeasurementLabels } from './MeasurementLabels';
 import { MeasurementLine } from './MeasurementLine';
 import { Measurement, MeasurementOptions } from './types';
+import { getNormalizedPixelCoordinates } from '@reveal/utilities/src/worldToViewport';
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply viewer/packages/tools/src/Measurement/MeasurementManager.ts]: 56171212b3db413ae1b60fe92250da930db103ac

File: viewer/packages/camera-manager/src/DefaultCameraManager.ts [hunk 3]
```diff
@@ -71,9 +78,11 @@ export class DefaultCameraManager implements CameraManager {
   private static readonly MouseDistanceThresholdBetweenRaycasts = 5;
   private static readonly DefaultCameraControlsOptions: Required<CameraControlsOptions> = {
     mouseWheelAction: 'zoomPastCursor',
-    changeCameraTargetOnClick: false
+    changeCameraTargetOnClick: false,
+    changeCameraPositionOnDoubleClick: false
   };
 
+  // NILS: Is this a copy contructor?
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply viewer/packages/camera-manager/src/DefaultCameraManager.ts:85]: Fixed

File: viewer/packages/camera-manager/src/ComboControls.ts [hunk 4]
```diff
@@ -929,147 +814,163 @@ export class ComboControls extends EventDispatcher<ComboControlsEventType> {
...
+    return this._vectors[this._index];
   }
 }
+
+//================================================
+// LOCAL CONSTANTS
+//================================================
+
+const IS_FIREFOX = navigator.userAgent.toLowerCase().indexOf('firefox') !== -1;
+const TARGET_FPS = 30;
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply viewer/packages/camera-manager/src/ComboControls.ts]: 1ff62b100bd161f1cd386c9c4b3bc5697a8c843f

File: viewer/packages/camera-manager/src/ComboControls.ts [hunk 5]
```diff
@@ -929,147 +814,163 @@ export class ComboControls extends EventDispatcher<ComboControlsEventType> {
...
-    domElement.removeEventListener('focus', this.onFocusChanged);
-    domElement.removeEventListener('blur', this.onFocusChanged);
+function isVectorAlmostZero(vector: Vector3, epsilon: number): boolean {
+  return Math.abs(vector.x) <= epsilon && Math.abs(vector.y) <= epsilon && Math.abs(vector.z) <= epsilon;
+}
 
-    window.removeEventListener('pointerup', this.onPointerUp);
-    window.removeEventListener('pointerdown', this.onFocusChanged);
+// Cache for using tempory vectors to avoid allocations
+class ReuseableVector3s {
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply viewer/packages/camera-manager/src/ComboControls.ts]: 1ff62b100bd161f1cd386c9c4b3bc5697a8c843f

File: viewer/packages/camera-manager/src/DefaultCameraManager.ts [hunk 6]
```diff
@@ -384,32 +390,34 @@ export class DefaultCameraManager implements CameraManager {
...
+        if (this._isDisposed) {
           return;
         }
-
         this.setComboControlsOptions({ lookAtViewTarget: false });
-        this.setComboControlsOptions({ enableKeyboardNavigation: true });
-
+        if (setKeyboardNavigationEnabled) {
+          this.keyboardNavigationEnabled = true;
+        }
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply viewer/packages/camera-manager/src/DefaultCameraManager.ts]: Easer now. The only problem is that fitCameraToBoundingBox is setting  keyboardNavigationEnabled to true. This was also the case before.

File: viewer/packages/camera-manager/src/ComboControls.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (greenbech) [reply viewer/packages/camera-manager/src/ComboControls.ts]: I can highly recommend this extension if you don't use it already! https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

**Pull request merged**